scene_synchronizer.cpp

```diff
@@ -2163,6 +2733,112 @@ void ClientSynchronizer::on_controller_reset(NetUtility::NodeData *p_node_data)
+
+	for (uint32_t g = 0; g < p_actions.size(); g += 1) {
+		const SenderNetAction &action = p_actions[g];
+
+		const bool already_received = server_sender_info.process_received_action(action.action_token);
+		if (already_received) {
+			// Already known nothing to do.
+			continue;
+		}
+
+		// Search the snapshot and add the Action to it.
+		// NOTE: This is needed in case of rewind to take the action into account.
+		NetworkedController *controller = static_cast<NetworkedController *>(player_controller_node_data->node);
+		const uint32_t current_input_id = controller->get_current_input_id();
+
+		if (action.client_get_executed_input_id() <= current_input_id) {
+			// On the server this action was executed on an already passed frame, insert it inside the snapshot.
+			// First search the snapshot:
+			bool add_to_snapshots = false;
+			for (uint32_t x = 0; x < client_snapshots.size(); x += 1) {
+				if (client_snapshots[x].input_id == action.client_get_executed_input_id()) {
+					// Insert the action inside the snapshot, so we can execute to reconcile the client and
+					// the server: I'm using `UINT32_MAX` because we track only locally executed actions.
+					client_snapshots[x].actions.push_back(TokenizedNetActionProcessor(UINT32_MAX, action.action_processor));
+					add_to_snapshots = true;
+					break;
+				}
+			}
+
+			if (!add_to_snapshots) {
+				NET_DEBUG_WARN("The Action `" + action.get_action_info().act_func + "` was not add to any snapshot as the snapshot was not found. The executed_input_id sent by the server is `" + itos(action.client_get_executed_input_id()) + "`. The actions is dropped.");
+				continue;
+			}
+		}
+
+		// Add the action to the pending actions so it's executed ASAP.
+		const uint32_t index = pending_actions.size();
+		pending_actions.push_back(action);
+		// Never networkd this back to server: afterall the server just sent it!
+		pending_actions[index].sent_by_the_server = true;
+
+		if (action.sender_executed_time_changed) {
+			// This action was generated by this peer, but it arrived too late to the server that
+			// rescheduled it:
+			// Remove the original Action from any stored snapshot to avoid executing it at the wrong time.
+			ERR_CONTINUE_MSG(action.triggerer_action_token == UINT32_MAX, "[FATAL] The server sent an action marked with `sender_executed_time_changed` but the `truggerer_action_token` is not set, this is a bug. Report that.");
+
+			for (uint32_t x = 0; x < client_snapshots.size(); x += 1) {
+				const int64_t action_i = client_snapshots[x].actions.find(TokenizedNetActionProcessor(action.triggerer_action_token, NetActionProcessor()));
+				if (action_i >= 0) {
+					client_snapshots[x].actions.remove(action_i);
+					break;
+				}
+			}
+
+			if (pending_actions[index].get_action_info().wait_server_validation == false) {
+				// Since it was already executed, no need to execute again, it's enough just put the Action
```

Comment:
Does this mean that all actions stored in a given snapshot will be executed again during rewind/replay?

Reply:
Yes, the action is executed again if a desync occurs and the client rewinds.